### DISCUSSION TITLE
Save Transformed Values of Node Attributes in MM File?

### DISCUSSION THREAD

**Discussion by WmLocke** [freeplane/freeplane#1941] [Ideas]
Hi 

Would it be possible to include a setting in Freeplane that saves the transformed values of node attributes as well their original values to the .mm file? This feature would enable processing these files with an XML tool, even when the attribute value is a formula. I guess there would have to be some compromise about the transformed values as they would all have to be evaluated as a string.

I understand that it's possible to write a Groovy script to export these transformed values, but this approach requires additional steps and reminders for colleagues to export the data whenever they update a map.

This setting would default to 'off' to maintain Freeplane's current behaviour.

Thank you for reading this request and for continuing to develop such a powerful and versatile tool.

Regards,

WmLocke

**Comment by macmarrum** [comment]
Could you please present an example, preferably an mind map?

I'd like to learn a bit more about the way you use Freeplane: is it formulas, formatting or numbering that are in the need of being transformed?

**Reply by WmLocke** [reply]
Thank you for replying.

I've attached a simplified version of the mind map we are creating.

I'm interested in the transformed results of formulas. Although we don't use value formats at the moment, it would be useful if transformed values saved to the .mm XML file respected these formats.

**Use Case for This Type of Map**

During the planning phase of a project, we move facility nodes around, and the moved node calculates which building, level, and room it occupies using attribute formulas. We also have attributes that sum the values of child nodes, such as Total Power and Total Data. Additionally, there are attributes with yes/no values indicating whether a particular facility, like a Staff Kitchen, exists somewhere in the lower parts of a branch.

The real maps are more complex and often have extra nodes at different levels to indicate queries, issues, and lessons learned.

**Benefits of Freeplane for Us**

Freeplane's conditional formatting at the map level, along with the ability to change the view root and filter nodes, allows us to quickly focus on different aspects of a project.

We have used other mind map tools in the past for workshops but quickly moved on to spreadsheets for more detailed planning. With Freeplane, we spend more time in the mind map, which helps with deliverables that are inherently hierarchical.

The downside of doing more of our planning in Freeplane is that much of the information is locked up in a map unless we write scripts to export it. While we can do this, it would scale better for our use cases if we could process the .mm files with XML tools to generate reports and tabular data for importing into other tools, like Excel. Hence, the desire for an option to automatically save transformed attribute values in the .mm file, ensuring the data is synchronised with the map's contents.

[Example Map With Attribute Formula.mm.zip](https://github.com/user-attachments/files/16542155/Example.Map.With.Attribute.Formula.mm.zip)

**Reply by macmarrum** [reply]
Thanks for the description and your mind map.

It's an interesting topic.

In addition to the idea presented by i-plasm, I can propose to save your mind map, with transformed values, to a JSON file. While I understand you're requirement is different, maybe JSON could be an option until you can get your transformed values in XML (.mm).

Here's a library I've been working on: https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/io/github/macmarrum/freeplane/Export.groovy

The attached mind map presents an example of how the library can be used. Please see `script1` on Root, which does an export to .json file and saves the json text as Root's note. Maybe it can help you process your data outside of Freeplane.

`script1` can be saved to a file and assigned a hot key `Ctrl+S` to replace regular Save, so that people don't have to remember to export json.

[macmarrum~Example Map With Attribute Formula.mm.zip](https://github.com/user-attachments/files/16567358/macmarrum.Example.Map.With.Attribute.Formula.mm.zip)

**Reply by WmLocke** [reply]
Hi macmarrum,

Thank you for your help. Overriding Ctrl+S is a good suggestion. I tried your Export library, and it worked well. We've been experimenting with different ways to process the JSON, and we’re leaning towards using a Python program to flatten the tree into a table in SQLite, which will allow us to run queries for reporting.

Regarding the Export library: have you considered adding support for attributes with duplicate names?

Currently, the script seems to output the value of the last instance of an attribute name. It might be useful to offer options for handling duplicates, such as:

- Returning the first instance
- Returning the last instance (current behaviour)
- Joining the values as a string with a delimiter
- Returning a list of values

The downside of the list option could be the potential for an inconsistent schema. For example, a single instance of an attribute would be output as a text string, but when duplicates are detected, it would be output as a JSON list. One possible solution is to pre-scan the map for duplicated attributes so that any attribute with duplicates is always output as a list, even when there’s only a single value.

That said, this might be too much of an edge case.

Thank you for creating such a useful library.

**Reply by macmarrum** [reply]
Hi WmLocke

I'm glad you are closer to finding a solution for your requirements.

> Regarding the Export library: have you considered adding support for attributes with duplicate names?

In the last 2 weeks, JSON support in Export/Import libraries has been improved, especially for attributes, which are now serialized as a map or a list of lists. The latter is used if there is more than one attribute with the same name or `settings.format == true`.
You might need to force-refresh the [Export.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/io/github/macmarrum/freeplane/Export.groovy) page (`Ctrl+F5`) to see the latest code.

Exmaples
```JSON
{
  "Duplicate-attribute-key example": {
    "@attributes": [
      ["vehicle", "bike"],
      ["vehicle", "car"],
    ]
  }
}
```
```json
{
  "Attributes-with-format example": {
    "@attributes": [
      ["vehicle", "bike"],
      ["date", "2024-08-25", "FormattedDate", "M/d yyyy"],
      ["number", 1.234, "FormattedNumber", "#0.0"]
    ]
  }
}
```
```json
{
  "Attributes-as-a-map example": {
    "@attributes": {
      "one": "eins",
      "two": 2
    }
  }
}
```


> We've been experimenting with different ways to process the JSON, and we’re leaning towards using a Python program to flatten the tree into a table in SQLite, which will allow us to run queries for reporting.

If you could open-source the solution and share it with the Freeplane community, it would be great.

**Reply by macmarrum** [reply]
> Regarding the Export library: have you considered adding support for attributes with duplicate names?
>> In the last 2 weeks, JSON support in Export/Import libraries has been improved, especially for attributes, which are now serialized as a map or a list of lists. The latter is used if there is more than one attribute with the same name or `settings.format == true`.

I've added a setting to force attributes to be serialized as a list of lists even if a map would suffice, so that the JSON output is always in the same format, making it easier to be processed → `settings = [forceAttribList: true]`

**Comment by i-plasm** [comment]
Hello @WmLocke  , 

It occurs to me that you can use a "evaluation script", set as a node attribute itself. The transformed values are assigned automatically on "eval" attributes (`_eval_<original_attr_name>`), so there's no need for any "extra" steps by the user. This video explains it better:


![evaluation_script](https://github.com/user-attachments/assets/67bde54b-e70e-4b83-81de-951f157113a8)


The evaluation script (i.e the `_eval_SCRIPT` attribute) needs to be included on every node, so you would have to employ some sort of system/workflow to ensure that. Ideally, such system would be automatic, not involving user action. It occurs to me Map Conditional Styles (of the script type) might be an alternative; there are probably others.

P.S The formula script used (*edit: wrapped now in `invokeLater(...)` and added an extra conditional because the previous version gave errors when the map was closed and reopened, complaining about a circular reference. More testing needs to be done but so far the error stopped occurring. But now I wonder if this approach has any unseen downsides*):

~~~ groovy
=
// Please include the "=" above for formulae
def prefix = "_eval_";
def attributes = node.attributes.transformed;

javax.swing.SwingUtilities.invokeLater{
    for (String name: attributes.names){
     if (!name.toLowerCase().startsWith(prefix)) {
        String val = attributes.find {it.key == name}.value;
        if (!attributes.get(name).equals(val)) {
            attributes.set(prefix + name, val);
        }
     }
    }
}
return "_SCRIPT_";
~~~

**Reply by i-plasm** [reply]
P.S The formula script used (*edit: wrapped now in `invokeLater(...)` and added an extra conditional because the previous version gave errors when the map was closed and reopened, complaining about a circular reference. More testing needs to be done but so far the error stopped occurring. But now I wonder if this approach has any unseen downsides*):

~~~ groovy
=
// Please include the "=" above for formulae
def prefix = "_eval_";
def attributes = node.attributes.transformed;

javax.swing.SwingUtilities.invokeLater{
    for (String name: attributes.names){
     if (!name.toLowerCase().startsWith(prefix)) {
        String val = attributes.find {it.key == name}.value;
        if (!attributes.get(name).equals(val)) {
            attributes.set(prefix + name, val);
        }
     }
    }
}
return "_SCRIPT_";
~~~

**Reply by WmLocke** [reply]
Hi I-plasm

Thank you. That's a very interesting solution. I think it might add too much visual clutter to the attributes panel. It's a pity FP doesn't allow for hidden attributes.

**Comment by Illegit** [comment]
I do not know if my note can help, but I added the ";" to separate the commands, and in doing so I avoided errors. Before the script was read as a unique line.... (my speech is surely not that of a programmer)
